### ISSUE TITLE
[mpi4py] Regression in RMA test

### ISSUE THREAD

**Issue by dalcinl** [pmodels/mpich#6292]
gforker + ch4:ofi: There is a RMA regression respect to about two weeks ago [[logs]](https://github.com/mpi4py/mpi4py-testing/actions/runs/3449138542/jobs/5835947548#step:17:1391)
```
testGetAccumulate (test_rma_nb.TestRMASelf) ... Assertion failed in file ./src/include/mpir_request.h at line 380: equal_ && (*(MPID_Thread_mutex_t *) MPIR_Request_mem[pool].lock).count >= 1
```

**Comment by dalcinl** [comment]
@hzhou I managed to find the commit introducing the regression

* Everything is OK up to commit 2e0db58f5c1ce60d6e3eb87e86b3b00368fedd00 (from #6250) [[logs]](https://github.com/mpi4py/mpi4py-testing/actions/runs/3499054867/jobs/5860168058).
* The singleton deadlocks and RMA failure both happen after 06a69645bdbf2837d014f0b6ddeb6548a7cd653b (from #6251) [[logs]](https://github.com/mpi4py/mpi4py-testing/actions/runs/3499265547/jobs/5860630126).

Therefore, looks like your change in ch4 to use default thread-cs to per-vci (perhaps 85a08849a04c8c09869b6f344d6362955bf26418 ?) caused the mpi4py testsuite regressions. 
As GitHub Actions do not run in bare metal (as opposed to MPICH's CI builders running Jenkins?), race conditions may trigger with ease, and that could explain why your Jenkins runs are good.

How should I proceed? Do you want me to fire a run using the global thread-cs with latest mpich@main? 
Can you remember me the exact configure option and value I should use?

**Comment by hzhou** [comment]
The failure is related to the per-vci lock. However, I don't think your test is a multithreaded test, right? So it is probably not a race condition. The assertion failure is likely saying a double locking condition. It should be easy to trace once I reproduce the failure...

**Comment by dalcinl** [comment]
You are right, most of my tests do not involve threads, and that's particularly the case of the failing RMA test. It should be a double-locking situation. I'll try to reproduce locally.

**Comment by dalcinl** [comment]
@hzhou I'm able to reproduce the issue locally. I manage to get a quick fix for `MPI_Rget_accumulate()` with runs using one MPI process (singleton or not), see https://github.com/dalcinl/mpich/commit/8ca02b7af5e248e728db85e9468837946bfa767c. The problem was not double-locking, but rater the lack or locking. Note my fix does not add the missing cs_enter/exit, but rather use a lightweight completed request as per the FIXME comment and related routines within the same file.
 
However, I still get a failure in `MPI_Raccumulate_c()` when using 2 MPI processes. To reproduce locally:
```
mpiexec -n 2 python test/main.py -i rma_nb -v -k '*World.testAccumulate'
```
I'm not able to debug further. Right now I do not have physical access to the workstation I'm running tests on, my network connectivity is really bad, I have to use a VPN, and I cannot manage to forward X11, so I don't know how to run gdb with two MPI processes. Sorry!

**Comment by hzhou** [comment]
@dalcinl Thank you for your debugging. That is very helpful. I'll investigate this issue after I wrap up another one that I am currently in the middle of. It is strange that our Jenkins didn't trigger it -- https://jenkins-pmrs.cels.anl.gov/job/mpich-main-mpi4py/337/ -- but good news that 32-bit is clean!

**Comment by dalcinl** [comment]
> It is strange that our Jenkins didn't trigger it

Please note I configure my builds with `--enable-g=all --enable-fast=none`. Are the mpi4py tests being run with an MPICH build configured with `--enable-g=all`? If not, that would explain why the assertion does not trigger in your Jenkins.

**Comment by hzhou** [comment]
Ahh, of course!
